### ISSUE TITLE
Performance improvements in kytheuri.Parse. [phab:D1522]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1937]
_Created by **fromberger** at 2017-04-14 00:59:47_: 

Add a simple parsing benchmark and use it to hammer on the performance of
parsing a bit. This change picks up some low-hanging fruit; there is probably
more we can squeeze if we want.

- Simplify scheme parsing (shortcut empty; don't loop).
- Separate Parse (with unescaping) from ParseRaw (structure).
- Split on strings.IndexByte rather than strings.Index.
- Streamline the unescaping code (drop bytes.Buffer).

With these changes, the runtime on a trivial benchmark drops by ~57%.

BEFORE

```
$ ./bazel-bin/kythe/go/util/kytheuri/kytheuri_test \
  -test.bench=. -test.count=4 -test.benchtime=5s
BenchmarkParse-6         3000000              2815 ns/op
BenchmarkParse-6         3000000              2807 ns/op
BenchmarkParse-6         3000000              2798 ns/op
BenchmarkParse-6         3000000              2800 ns/op
```

Mean: 2805 ns/op

AFTER

```
$ ./bazel-bin/kythe/go/util/kytheuri/kytheuri_test \
  -test.bench=. -test.count=4 -test.benchtime=5s
BenchmarkParse-6        10000000              1193 ns/op
BenchmarkParse-6        10000000              1185 ns/op
BenchmarkParse-6        10000000              1177 ns/op
BenchmarkParse-6        10000000              1187 ns/op
```

Mean: 1185 ns/op

AFTER (take 2)

```
$ ./bazel-bin/kythe/go/util/kytheuri/kytheuri_test
  -test.bench=. -test.count=4 -test.benchtime=5s
BenchmarkParse-6        10000000               993 ns/op
BenchmarkParse-6        10000000              1001 ns/op
BenchmarkParse-6        10000000               997 ns/op
BenchmarkParse-6        10000000               982 ns/op
```

Mean: 993 ns/op

AFTER (take 3)

```
$ ./bazel-bin/kythe/go/util/kytheuri/kytheuri_test \
  -test.bench=. -test.count=4 -test.benchtime=5s 
BenchmarkParse-6        10000000               693 ns/op
BenchmarkParse-6        10000000               665 ns/op
BenchmarkParse-6        10000000               664 ns/op
BenchmarkParse-6        10000000               668 ns/op
```

Mean: 672 ns/op

Runtime is now ~76% lower than originally.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-14 05:59:33_:

More improvements from CPU profiling.

  - Replace another strings.Index by strings.IndexByte.
  - Reuse a buffer for unescaping.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-14 06:48:19_:

Another hit from profiling:

 - Use a custom splitter for attributes.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-04-14 16:15:19_:

It might be worth separating the return type since the URI cannot have String called on it normally.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-14 17:23:31_:

  - Make a separate type for raw (escaped) Kythe URIs.

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2017-04-14 17:44:20_

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-14 17:47:20_:

Good idea. I did that. I also added methods to go back and forth between the two representations, and pushed the string-encoding up a level.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-14 17:47:21_:

  - kytheuri: Add an Encode method to the *URI type.

**Comment by kythe** [comment]
_**schroederc** wrote at 2017-04-14 17:59:43_:

LGTM

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-14 19:12:53_:

  - Export the URI wrapped in a *Raw value.

**Comment by kythe** [comment]
_**fromberger** closed via commit 4bf509757e0ccf50d719f98e4a2b616be5efff3c at 2017-04-17 12:08:19_
